File: src/main/java/de/fhopf/akka/HtmlParserPageRetriever.java
package de.fhopf.akka;

import java.util.ArrayList;
import java.util.List;
import org.htmlparser.Parser;
import org.htmlparser.Tag;
import org.htmlparser.tags.BodyTag;
import org.htmlparser.tags.LinkTag;
import org.htmlparser.tags.TitleTag;
import org.htmlparser.util.ParserException;
import org.htmlparser.visitors.NodeVisitor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Retrieves the content using HtmlParser.
 *
 * @author Florian Hopf, http://www.florian-hopf.de
 */
public class HtmlParserPageRetriever implements PageRetriever {

    private final String baseUrl;
    private static final Logger logger = LoggerFactory.getLogger(HtmlParserPageRetriever.class);

    public HtmlParserPageRetriever(String baseUrl) {
        this.baseUrl = baseUrl;
    }

    @Override
    public PageContent fetchPageContent(String url) {
        logger.debug("Fetching {}", url);
        try {
            Parser parser = new Parser(url);
            PageContentVisitor visitor = new PageContentVisitor(baseUrl, url);
            parser.visitAllNodesWith(visitor);
            
            return visitor.getContent();
        } catch (ParserException ex) {
            throw new IllegalStateException(ex);
        }
    }


    private static class PageContentVisitor extends NodeVisitor {

        private List<String> linksToVisit = new ArrayList<String>();
        private String content;
        private String title;

        private final String baseUrl;
        private final String currentUrl;
        
        public PageContentVisitor(String baseUrl, String currentUrl) {
            super(true);
            this.baseUrl = baseUrl;
            this.currentUrl = currentUrl;
            
        }

        @Override
        public void visitTag(Tag tag) {
            if (tag instanceof LinkTag) {
                LinkTag linkTag = (LinkTag) tag;
                if (linkTag.getLink().startsWith(baseUrl) && isProbablyHtml(linkTag.getLink())) {
                    logger.debug("Using link pointing to {}", linkTag.getLink());
                    linksToVisit.add(linkTag.getLink());
                } else {
                    logger.debug("Skipping link pointing to {}", linkTag.getLink());
                }
            } else if (tag instanceof TitleTag) {
                TitleTag titleTag = (TitleTag) tag;
                title = titleTag.getTitle();
            } else if (tag instanceof BodyTag) {
                BodyTag bodyTag = (BodyTag) tag;
                content = bodyTag.toPlainTextString();
            }
        }
        
        public PageContent getContent() {
            return new PageContent(currentUrl, linksToVisit, title, content);
        }

        private boolean isProbablyHtml(String link) {
            return link.endsWith(".html") || link.endsWith("/");
        }
    }
}

File: src/main/scala/de/fhopf/akka/HtmlParserPageRetriever.scala
package de.fhopf.akka

import java.util.ArrayList
import java.util.List
import org.htmlparser.Parser
import org.htmlparser.Tag
import org.htmlparser.tags.BodyTag
import org.htmlparser.tags.LinkTag
import org.htmlparser.tags.TitleTag
import org.htmlparser.util.ParserException
import org.htmlparser.visitors.NodeVisitor
import org.slf4j.Logger
import org.slf4j.LoggerFactory

/**
 * @author Florian Hopf, http://www.florian-hopf.de
 * @author Alfredo Serafini, http://www.seralf.it
 */
class HtmlParserPageRetriever(baseUrl: String) extends PageRetriever {

  private val logger: Logger = LoggerFactory.getLogger(classOf[HtmlParserPageRetriever])

  override def fetchPageContent(url: String): PageContent = {
    logger.debug("Fetching {}", url)
    try {
      val parser: Parser = new Parser(url)
      val visitor: PageContentVisitor = new PageContentVisitor(baseUrl, url)
      parser.visitAllNodesWith(visitor)
      visitor.getContent()
    } catch {
      case ex: ParserException => throw new IllegalStateException(ex)
    }
  }

  class PageContentVisitor(recursive: Boolean) extends NodeVisitor(recursive) {

    private val linksToVisit: List[String] = new ArrayList[String]()
    private var content: String = ""
    private var title: String = ""

    private var baseUrl: String = ""
    private var currentUrl: String = ""

    def this(baseUrl: String, currentUrl: String) {
      this(true)
      this.baseUrl = baseUrl
      this.currentUrl = currentUrl

    }

    override def visitTag(tag: Tag) = {
      tag match {
        case linkTag: LinkTag => {
          if (linkTag.getLink().startsWith(baseUrl) && isProbablyHtml(linkTag.getLink())) {
            logger.debug("Using link pointing to {}", linkTag.getLink())
            linksToVisit.add(linkTag.getLink())
          } else {
            logger.debug("Skipping link pointing to {}", linkTag.getLink())
          }
        }
        case titleTag: TitleTag => {
          title = titleTag.getTitle()
        }
        case bodyTag: BodyTag => {
          content = bodyTag.toPlainTextString()
        }
      }
    }

    def getContent(): PageContent = new PageContent(currentUrl, linksToVisit, title, content)

    private def isProbablyHtml(link: String): Boolean = link.endsWith(".html") || link.endsWith("/")
  }
}
